searchlib/src/vespa/searchlib/features/internal_max_reduce_prod_join_feature.cpp

```diff
@@ -0,0 +1,226 @@
+    RawExecutor<BaseType>::outputs().set_number(0, val);
+}
+
+/**
+ * Blueprint
+ */
+InternalMaxReduceProdJoinBlueprint::InternalMaxReduceProdJoinBlueprint() :
+        Blueprint("internalMaxReduceProdJoin")
+{
+}
+
+InternalMaxReduceProdJoinBlueprint::~InternalMaxReduceProdJoinBlueprint()
+{
+}
+
+void
+InternalMaxReduceProdJoinBlueprint::visitDumpFeatures(const IIndexEnvironment &,
+                                                           IDumpFeatureVisitor &) const
+{
+}
+
+Blueprint::UP
+InternalMaxReduceProdJoinBlueprint::createInstance() const
+{
+    return Blueprint::UP(new InternalMaxReduceProdJoinBlueprint());
+}
+
+ParameterDescriptions
+InternalMaxReduceProdJoinBlueprint::getDescriptions() const
+{
+    return ParameterDescriptions().desc().attribute(ParameterCollection::ARRAY).string();
+}
+
+bool
+InternalMaxReduceProdJoinBlueprint::setup(const IIndexEnvironment &env, const ParameterList &params)
+{
+    _attribute = params[0].getValue();
+    _query = params[1].getValue();
+    describeOutput("scalar", "Internal executor for optimized execution of reduce(join(A,Q,f(x,y)(x*y)),max)");
+    env.hintAttributeAccess(_attribute);
+    return true;
+}
+
+bool isImportedAttribute(const IAttributeVector& attribute) noexcept {
+    return dynamic_cast<const ImportedAttributeVector*>(&attribute) != nullptr;
+}
+
+template<typename A>
+bool supportsGetRawValues(const A &attr) noexcept {
+    try {
+        const multivalue::Value<typename A::BaseType> *tmp = nullptr;
+        attr.getRawValues(0, tmp); // Throws if unsupported
+        return true;
+    } catch (const std::runtime_error &e) {
+        (void) e;
+        return false;
+    }
+}
+
+template <typename BaseType>
+FeatureExecutor &
+selectTypedExecutor(const IAttributeVector* attribute, const IntegerVector& vector, vespalib::Stash &stash)
+{
+    if (!isImportedAttribute(*attribute)) {
+        using A = IntegerAttributeTemplate<BaseType>;
+        using VT = multivalue::Value<BaseType>;
+        using ExactA = MultiValueNumericAttribute<A, VT>;
+
+        const A *iattr = dynamic_cast<const A *>(attribute);
+        if (supportsGetRawValues(*iattr)) {
+            const ExactA *exactA = dynamic_cast<const ExactA *>(iattr);
+            if (exactA != nullptr) {
+                return stash.create<RawExecutor<BaseType>>(attribute, vector);
+            }
+        }
+    }
+    return stash.create<BufferedExecutor<BaseType>>(attribute, vector);
+}
+
+FeatureExecutor &
+selectExecutor(const IAttributeVector* attribute, const IntegerVector& vector, vespalib::Stash &stash)
+{
+    if (attribute->getCollectionType() == CollectionType::ARRAY) {
+        switch (attribute->getBasicType()) {
+            case BasicType::INT32:
+                return selectTypedExecutor<int32_t>(attribute, vector, stash);
+            case BasicType::INT64:
+                return selectTypedExecutor<int64_t>(attribute, vector, stash);
+            default:
+                break;
```

Comment:
if the cases not supported here are not rejected in setup, they will stop working when expression replacement is enabled later.